test/unit/trigger.c

```diff
@@ -0,0 +1,202 @@
+ */
+enum func_type {
+	/** Do nothing. */
+	FUNC_TYPE_NOP,
+	/** Error. */
+	FUNC_TYPE_ERR,
+	/** Swap trigger list heads. */
+	FUNC_TYPE_SWAP,
+	/** Clear one of the triggers: self or other. */
+	FUNC_TYPE_CLEAR,
+};
+
+static int
+func_type_by_no(int func_no)
+{
+	assert(func_no >= 0 && func_no < FUNC_COUNT);
+	if (func_no < TEST_LENGTH)
+		return FUNC_TYPE_CLEAR;
+	else if (func_no == TEST_LENGTH)
+		return FUNC_TYPE_ERR;
+	else if (func_no == TEST_LENGTH + 1)
+		return FUNC_TYPE_NOP;
+	else /* if (func_no == TEST_LENGTH + 2) */
+		return FUNC_TYPE_SWAP;
+}
+
+static void
+fill_trigger_list(struct rlist *list, bool *should_run, int direction)
+{
+	for (int i = 0; i < TEST_LENGTH; i++)
+		should_run[i] = true;
+	rlist_create(list);
+	/*
+	 * Create triggers and deduce which should and which shouldn't run based
+	 * on trigger functions and run direction.
+	 */
+	for (int i = (TEST_LENGTH - 1) * (1 - direction) / 2;
+	     i >= 0 && i < TEST_LENGTH; i += direction) {
+		was_run[i] = 0;
+		triggers[i].id = i;
+		int func_no = funcs[i];
+		switch (func_type_by_no(func_no)) {
+		case FUNC_TYPE_CLEAR:
+			trigger_create(&triggers[i].base,
+				       trigger_clear_f, NULL, NULL);
+			triggers[i].target_id = func_no;
+			if (!should_run[i])
+				break;
+			int target_id = funcs[i];
+			if (direction * i < direction * target_id)
+				should_run[target_id] = false;
+			break;
+		case FUNC_TYPE_ERR:
+			trigger_create(&triggers[i].base, trigger_err_f,
+				       NULL, NULL);
+			if (!should_run[i])
+				break;
+			for (int j = i + direction; j >= 0 && j < TEST_LENGTH;
+			     j += direction) {
+				should_run[j] = false;
+			}
+			break;
+		case FUNC_TYPE_NOP:
+			trigger_create(&triggers[i].base, trigger_nop_f,
+				       NULL, NULL);
+			break;
+		case FUNC_TYPE_SWAP:
+			trigger_create(&triggers[i].base,
+				       trigger_swap_f, NULL, NULL);
+			break;
+		}
+	}
+	/*
+	 * Add triggers in reverse order, so that trigger[0] is first to run in
+	 * direct order and last to run in reverse order.
+	 */
+	for (int i = TEST_LENGTH - 1; i >= 0; i--)
+		trigger_add(list, &triggers[i].base);
+}
+
+static void
+test_trigger_one(void)
+{
+	plan(2 * TEST_LENGTH);
+	for (int direction = -1; direction <= 1; direction += 2) {
+		bool should_run[TEST_LENGTH];
+		fill_trigger_list(&list_a, should_run, direction);
+		if (direction == 1)
+			trigger_run(&list_a, NULL);
+		else
+			trigger_run_reverse(&list_a, NULL);
+		for (int i = 0; i < TEST_LENGTH; i++) {
+			ok((should_run[i] && was_run[i] == 1) ||
+			   (!should_run[i] && was_run[i] == 0),
+			   "Triggers ran correctly");
+		}
+	}
+	check_plan();
+}
+
+static void
+test_trigger(int pos, bool had_clear)
+{
+	if (pos == TEST_LENGTH)
+		return test_trigger_one();
+	int i = had_clear ? TEST_LENGTH : 0;
+	plan(FUNC_COUNT - i);
+	for (i; i < FUNC_COUNT; i++) {
+		funcs[pos] = i;
+		test_trigger(pos + 1, had_clear || func_type_by_no(i) ==
+						   FUNC_TYPE_CLEAR);
+	}
```

Comment:
I struggle to understand the test. Can you rework it aiming at readability instead of code reuse? Since it's just a test, code duplication is acceptable. What I'd like to see is a set of independent tests, each of which checks a particular use case:
1. Deletion of self from a trigger callback.
2. Deletion of the next/prev trigger.
3. Swap of the trigger list head.
4. ...

Reply:
I would rather leave this test, because it checks for some unexpected trigger combinations we wouldn't think of ourselves: like:
1. clearing the next/prev trigger and swapping list head
2. swapping list head and having an errored trigger
3. so on

I don't see how these particular cases could fail now, but they may be broken by some future patches.

IOW cases you propose only cover "bad" triggers one by one, while I propose to cover all their combinations.

Now how the test works:
It iterates over all possible combinations of triggers of length 5, runs them all in direct and reverse order and checks whether what should run is run and what shouldn't run isn't run.

Possible triggers to choose from are (for each position):
1. Clear trigger at position i, i=1..5
2. Swap list head
3. Nothing bad
4. Error

I don't really check **each** possible combination, because it would make the test quite time-consuming. For example, if there is one trigger clearing something, I don't add any more clearing triggers (checking all combinations of clear(1..5) for all the positions would require a lot more iterations).
I chose trigger lists of length 5, because 5 is the first length, where there is a trigger for which clearing self or next or prev isn't the same as clearing first or last trigger (which might also be a special case).j

I can try and rewrite it to be more understandable or add some comments.

Reply:
The problem is the test is way too complicated IMO. In fact, its code is more complicated than the code of the patch itself. If we find another bug in the code, it'll take us a substantial amount of time to extend the test. We'll probably just give up and add a simple independent test case alongside. That's why I can't approve of it. However, I won't block it either if @Gerold103 reviews and approves it.